File: gemini/typescript/examples/embeddings.ts
/**
 * Embeddings Example
 *
 * This example demonstrates:
 * - Generating single embeddings
 * - Batch embedding generation
 * - Different task types (retrieval, classification, etc.)
 * - Using embeddings for semantic similarity
 *
 * Prerequisites:
 * - Set the GEMINI_API_KEY environment variable
 *
 * Usage:
 * ```bash
 * export GEMINI_API_KEY="your-api-key"
 * npm run build
 * node dist/examples/embeddings.js
 * ```
 */

import { createClientFromEnv } from '../src/index.js';

/**
 * Calculate cosine similarity between two vectors
 */
function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) {
    throw new Error('Vectors must have the same length');
  }

  let dotProduct = 0;
  let magnitudeA = 0;
  let magnitudeB = 0;

  for (let i = 0; i < a.length; i++) {
    dotProduct += a[i] * b[i];
    magnitudeA += a[i] * a[i];
    magnitudeB += b[i] * b[i];
  }

  magnitudeA = Math.sqrt(magnitudeA);
  magnitudeB = Math.sqrt(magnitudeB);

  if (magnitudeA === 0 || magnitudeB === 0) {
    return 0;
  }

  return dotProduct / (magnitudeA * magnitudeB);
}

/**
 * Example 1: Generate a single embedding
 */
async function exampleSingleEmbedding(): Promise<number[]> {
  console.log('\n=== Example 1: Single Embedding ===\n');

  const client = createClientFromEnv();

  const text = 'What is the capital of France?';
  console.log(`Generating embedding for: "${text}"\n`);

  const response = await client.embeddings.embed('models/text-embedding-004', {
    content: {
      parts: [{ text }],
    },
    taskType: 'RETRIEVAL_QUERY',
  });

  console.log('Embedding generated successfully!');
  console.log(`Dimensions: ${response.embedding.values.length}`);
  console.log(`First 5 values: [${response.embedding.values.slice(0, 5).join(', ')}...]`);

  return response.embedding.values;
}

/**
 * Example 2: Batch embeddings
 */
async function exampleBatchEmbeddings(): Promise<number[][]> {
  console.log('\n=== Example 2: Batch Embeddings ===\n');

  const client = createClientFromEnv();

  const texts = [
    'Paris is the capital of France.',
    'London is the capital of the United Kingdom.',
    'Berlin is the capital of Germany.',
    'Madrid is the capital of Spain.',
    'Rome is the capital of Italy.',
  ];

  console.log('Generating embeddings for multiple texts...');
  console.log(`Number of texts: ${texts.length}\n`);

  const requests = texts.map((text) => ({
    content: {
      parts: [{ text }],
    },
    taskType: 'RETRIEVAL_DOCUMENT' as const,
  }));

  const response = await client.embeddings.batchEmbed('models/text-embedding-004', requests);

  console.log('Batch embeddings generated successfully!');
  console.log(`Number of embeddings: ${response.embeddings.length}`);

  if (response.embeddings.length > 0) {
    console.log(`Embedding dimensions: ${response.embeddings[0].values.length}`);
  }

  // Show first few values of each embedding
  console.log('\nFirst 3 values of each embedding:');
  for (let i = 0; i < response.embeddings.length; i++) {
    const values = response.embeddings[i].values;
    console.log(`${i + 1}. [${values.slice(0, 3).join(', ')}...]`);
  }

  return response.embeddings.map((emb) => emb.values);
}

/**
 * Example 3: Different task types
 */
async function exampleTaskTypes(): Promise<void> {
  console.log('\n=== Example 3: Different Task Types ===\n');

  const client = createClientFromEnv();

  const text = 'Machine learning is a subset of artificial intelligence.';

  // Task types available:
  // - RETRIEVAL_QUERY: For search queries
  // - RETRIEVAL_DOCUMENT: For documents in a corpus
  // - SEMANTIC_SIMILARITY: For similarity comparison
  // - CLASSIFICATION: For text classification
  // - CLUSTERING: For clustering similar texts

  const taskTypes = [
    'RETRIEVAL_QUERY',
    'RETRIEVAL_DOCUMENT',
    'SEMANTIC_SIMILARITY',
    'CLASSIFICATION',
    'CLUSTERING',
  ] as const;

  console.log(`Generating embeddings with different task types for:`);
  console.log(`"${text}"\n`);

  for (const taskType of taskTypes) {
    const response = await client.embeddings.embed('models/text-embedding-004', {
      content: {
        parts: [{ text }],
      },
      taskType,
    });

    console.log(`Task Type: ${taskType}`);
    console.log(`  Dimensions: ${response.embedding.values.length}`);
    console.log(`  First 3 values: [${response.embedding.values.slice(0, 3).join(', ')}...]`);
    console.log();
  }
}

/**
 * Example 4: Semantic similarity comparison
 */
async function exampleSemanticSimilarity(): Promise<void> {
  console.log('\n=== Example 4: Semantic Similarity ===\n');

  const client = createClientFromEnv();

  // Base query
  const query = 'What is the capital city of France?';

  // Similar and dissimilar sentences
  const sentences = [
    'Paris is the capital of France.',
    'The capital of France is Paris.',
    'France has Paris as its capital city.',
    'I love eating pizza for dinner.',
    'The weather is sunny today.',
  ];

  console.log(`Query: "${query}"\n`);
  console.log('Comparing with sentences:\n');

  // Generate embedding for query
  const queryResponse = await client.embeddings.embed('models/text-embedding-004', {
    content: {
      parts: [{ text: query }],
    },
    taskType: 'RETRIEVAL_QUERY',
  });

  const queryEmbedding = queryResponse.embedding.values;

  // Generate embeddings for sentences
  const sentenceRequests = sentences.map((text) => ({
    content: {
      parts: [{ text }],
    },
    taskType: 'RETRIEVAL_DOCUMENT' as const,
  }));

  const sentencesResponse = await client.embeddings.batchEmbed(
    'models/text-embedding-004',
    sentenceRequests
  );

  // Calculate similarities
  console.log('Similarity scores:\n');

  const similarities = sentencesResponse.embeddings.map((emb, index) => ({
    sentence: sentences[index],
    similarity: cosineSimilarity(queryEmbedding, emb.values),
  }));

  // Sort by similarity (descending)
  similarities.sort((a, b) => b.similarity - a.similarity);

  for (const { sentence, similarity } of similarities) {
    const percentage = (similarity * 100).toFixed(2);
    const bar = '█'.repeat(Math.floor(similarity * 50));
    console.log(`${percentage}% ${bar}`);
    console.log(`  "${sentence}"`);
    console.log();
  }
}

/**
 * Example 5: Text clustering
 */
async function exampleClustering(): Promise<void> {
  console.log('\n=== Example 5: Text Clustering ===\n');

  const client = createClientFromEnv();

  // Documents on different topics
  const documents = [
    // Technology
    'Artificial intelligence is transforming many industries.',
    'Machine learning models require large datasets for training.',
    'Neural networks are inspired by biological brain structure.',
    // Cooking
    'Baking bread requires flour, water, yeast, and salt.',
    'Italian cuisine is known for pasta, pizza, and risotto.',
    'Proper knife skills are essential for any chef.',
    // Sports
    'Basketball is played with two teams of five players.',
    'Soccer is the most popular sport in the world.',
    'Tennis matches can last for several hours.',
  ];

  console.log('Generating embeddings for clustering...');
  console.log(`Number of documents: ${documents.length}\n`);

  const requests = documents.map((text) => ({
    content: {
      parts: [{ text }],
    },
    taskType: 'CLUSTERING' as const,
  }));

  const response = await client.embeddings.batchEmbed('models/text-embedding-004', requests);

  console.log('Computing pairwise similarities...\n');

  // Compute similarity matrix
  const embeddings = response.embeddings.map((emb) => emb.values);

  console.log('Documents grouped by similarity:\n');

  // Simple clustering: find most similar documents
  const processed = new Set<number>();

  for (let i = 0; i < documents.length; i++) {
    if (processed.has(i)) continue;

    const cluster = [i];
    processed.add(i);

    for (let j = i + 1; j < documents.length; j++) {
      if (processed.has(j)) continue;

      const similarity = cosineSimilarity(embeddings[i], embeddings[j]);
      if (similarity > 0.6) {
        // Threshold for clustering
        cluster.push(j);
        processed.add(j);
      }
    }

    console.log(`Cluster ${processed.size / cluster.length}:`);
    for (const idx of cluster) {
      console.log(`  - ${documents[idx]}`);
    }
    console.log();
  }
}

/**
 * Example 6: Using embeddings with titles
 */
async function exampleWithTitle(): Promise<void> {
  console.log('\n=== Example 6: Embeddings with Titles ===\n');

  const client = createClientFromEnv();

  const title = 'Introduction to Quantum Computing';
  const content = `Quantum computing uses quantum mechanical phenomena such as superposition
and entanglement to process information in ways that classical computers cannot.`;

  console.log(`Title: "${title}"`);
  console.log(`Content: "${content}"\n`);

  const response = await client.embeddings.embed('models/text-embedding-004', {
    content: {
      parts: [{ text: content }],
    },
    taskType: 'RETRIEVAL_DOCUMENT',
    title,
  });

  console.log('Embedding with title generated successfully!');
  console.log(`Dimensions: ${response.embedding.values.length}`);
  console.log(`First 5 values: [${response.embedding.values.slice(0, 5).join(', ')}...]`);
}

/**
 * Main function running all examples
 */
async function main(): Promise<void> {
  try {
    console.log('=== Embeddings Examples ===');

    await exampleSingleEmbedding();
    await exampleBatchEmbeddings();
    await exampleTaskTypes();
    await exampleSemanticSimilarity();
    await exampleClustering();
    await exampleWithTitle();

    console.log('\n=== All examples completed successfully! ===');
  } catch (error) {
    console.error('\nError during embeddings generation:', error);
    if (error instanceof Error) {
      console.error('Error message:', error.message);
      console.error('Stack trace:', error.stack);
    }
    process.exit(1);
  }
}

// Run the examples
main();

File: gemini/rust/examples/embeddings.rs
//! Embeddings generation example for Gemini API.
//!
//! This example demonstrates:
//! - Generating single embeddings
//! - Batch embedding generation
//! - Different task types (RETRIEVAL_QUERY, RETRIEVAL_DOCUMENT, etc.)
//! - Understanding embedding dimensions
//! - Using embeddings for similarity search
//!
//! # Usage
//!
//! Set your API key as an environment variable:
//! ```bash
//! export GEMINI_API_KEY="your-api-key-here"
//! ```
//!
//! Then run:
//! ```bash
//! cargo run --example embeddings
//! ```

use integrations_gemini::{
    GeminiClientImpl,
    types::{
        EmbedContentRequest, Content, Part, TaskType,
    },
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::INFO)
        .with_target(false)
        .init();

    println!("=== Gemini Embeddings Example ===\n");

    // Create client
    println!("1. Creating Gemini client from environment...");
    let client = GeminiClientImpl::from_env()?;
    println!("   ✓ Client created successfully\n");

    // Example 1: Single embedding
    println!("=== Example 1: Single Text Embedding ===\n");
    example_single_embedding(&client).await?;

    println!("\n");

    // Example 2: Batch embeddings
    println!("=== Example 2: Batch Embeddings ===\n");
    example_batch_embeddings(&client).await?;

    println!("\n");

    // Example 3: Different task types
    println!("=== Example 3: Task Types ===\n");
    example_task_types(&client).await?;

    println!("\n");

    // Example 4: Semantic similarity
    println!("=== Example 4: Semantic Similarity ===\n");
    example_semantic_similarity(&client).await?;

    println!("\n=== Examples Complete ===");

    Ok(())
}

/// Example: Generate a single embedding.
async fn example_single_embedding(client: &GeminiClientImpl) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Generating single embedding...");

    let text = "The quick brown fox jumps over the lazy dog.";
    println!("   Text: {}", text);

    let request = EmbedContentRequest {
        content: Content {
            role: None,
            parts: vec![
                Part::Text {
                    text: text.to_string(),
                }
            ],
        },
        task_type: Some(TaskType::RetrievalDocument),
        title: None,
        output_dimensionality: None,
    };

    println!("   Task type: RETRIEVAL_DOCUMENT");
    println!("   Model: text-embedding-004");

    // Note: EmbeddingsService is implemented
    let response = client.embeddings()
        .embed("text-embedding-004", request)
        .await?;

    println!("\n   ✓ Embedding generated successfully");

    if let Some(values) = &response.embedding.values {
        println!("   Embedding dimension: {}", values.len());
        println!("   First 5 values: {:?}", &values[..5.min(values.len())]);

        // Calculate L2 norm
        let norm: f32 = values.iter().map(|v| v * v).sum::<f32>().sqrt();
        println!("   L2 norm: {:.6}", norm);
    }

    Ok(())
}

/// Example: Generate batch embeddings.
async fn example_batch_embeddings(client: &GeminiClientImpl) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Generating batch embeddings...");

    let texts = vec![
        "Machine learning is a subset of artificial intelligence.",
        "Deep learning uses neural networks with many layers.",
        "Natural language processing helps computers understand text.",
        "Computer vision enables machines to interpret images.",
    ];

    println!("   Generating embeddings for {} texts", texts.len());

    let requests: Vec<EmbedContentRequest> = texts.iter().map(|text| {
        EmbedContentRequest {
            content: Content {
                role: None,
                parts: vec![
                    Part::Text {
                        text: text.to_string(),
                    }
                ],
            },
            task_type: Some(TaskType::RetrievalDocument),
            title: None,
            output_dimensionality: None,
        }
    }).collect();

    // Note: EmbeddingsService is implemented
    let response = client.embeddings()
        .batch_embed("text-embedding-004", requests)
        .await?;

    println!("\n   ✓ Batch embeddings generated successfully");

    if let Some(embeddings) = &response.embeddings {
        println!("   Generated {} embeddings", embeddings.len());

        for (i, embedding) in embeddings.iter().enumerate() {
            if let Some(values) = &embedding.values {
                println!("\n   Embedding {}:", i + 1);
                println!("     Text: {}", &texts[i]);
                println!("     Dimension: {}", values.len());
                println!("     First 3 values: {:?}", &values[..3.min(values.len())]);
            }
        }
    }

    Ok(())
}

/// Example: Different task types.
async fn example_task_types(client: &GeminiClientImpl) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Demonstrating different task types...");

    let text = "How do I create embeddings with Gemini?";

    // Example: Query embedding
    println!("\n   Task Type: RETRIEVAL_QUERY");
    println!("   Use case: Search queries, questions");

    let query_request = EmbedContentRequest {
        content: Content {
            role: None,
            parts: vec![
                Part::Text {
                    text: text.to_string(),
                }
            ],
        },
        task_type: Some(TaskType::RetrievalQuery),
        title: None,
        output_dimensionality: None,
    };

    let query_response = client.embeddings()
        .embed("text-embedding-004", query_request)
        .await?;

    if let Some(values) = &query_response.embedding.values {
        println!("   ✓ Query embedding generated: {} dimensions", values.len());
    }

    // Example: Document embedding
    println!("\n   Task Type: RETRIEVAL_DOCUMENT");
    println!("   Use case: Documents to be searched, knowledge base entries");

    let doc_request = EmbedContentRequest {
        content: Content {
            role: None,
            parts: vec![
                Part::Text {
                    text: "Embeddings are vector representations of text.".to_string(),
                }
            ],
        },
        task_type: Some(TaskType::RetrievalDocument),
        title: Some("Introduction to Embeddings".to_string()),
        output_dimensionality: None,
    };

    let doc_response = client.embeddings()
        .embed("text-embedding-004", doc_request)
        .await?;

    if let Some(values) = &doc_response.embedding.values {
        println!("   ✓ Document embedding generated: {} dimensions", values.len());
    }

    // Example: Semantic similarity
    println!("\n   Task Type: SEMANTIC_SIMILARITY");
    println!("   Use case: Comparing text similarity, clustering");

    let sim_request = EmbedContentRequest {
        content: Content {
            role: None,
            parts: vec![
                Part::Text {
                    text: "Text embedding for similarity comparison.".to_string(),
                }
            ],
        },
        task_type: Some(TaskType::SemanticSimilarity),
        title: None,
        output_dimensionality: None,
    };

    let sim_response = client.embeddings()
        .embed("text-embedding-004", sim_request)
        .await?;

    if let Some(values) = &sim_response.embedding.values {
        println!("   ✓ Similarity embedding generated: {} dimensions", values.len());
    }

    // Example: Classification
    println!("\n   Task Type: CLASSIFICATION");
    println!("   Use case: Text categorization, sentiment analysis");

    let class_request = EmbedContentRequest {
        content: Content {
            role: None,
            parts: vec![
                Part::Text {
                    text: "This product is amazing!".to_string(),
                }
            ],
        },
        task_type: Some(TaskType::Classification),
        title: None,
        output_dimensionality: None,
    };

    let class_response = client.embeddings()
        .embed("text-embedding-004", class_request)
        .await?;

    if let Some(values) = &class_response.embedding.values {
        println!("   ✓ Classification embedding generated: {} dimensions", values.len());
    }

    // Example: Clustering
    println!("\n   Task Type: CLUSTERING");
    println!("   Use case: Grouping similar documents, topic modeling");

    let cluster_request = EmbedContentRequest {
        content: Content {
            role: None,
            parts: vec![
                Part::Text {
                    text: "Document for clustering analysis.".to_string(),
                }
            ],
        },
        task_type: Some(TaskType::Clustering),
        title: None,
        output_dimensionality: None,
    };

    let cluster_response = client.embeddings()
        .embed("text-embedding-004", cluster_request)
        .await?;

    if let Some(values) = &cluster_response.embedding.values {
        println!("   ✓ Clustering embedding generated: {} dimensions", values.len());
    }

    println!("\n   Note: Different task types optimize embeddings for specific use cases.");

    Ok(())
}

/// Example: Computing semantic similarity.
async fn example_semantic_similarity(client: &GeminiClientImpl) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Computing semantic similarity...");

    let texts = vec![
        "The cat sat on the mat.",
        "A feline rested on the rug.",
        "The dog played in the park.",
    ];

    println!("\n   Comparing similarity between:");
    for (i, text) in texts.iter().enumerate() {
        println!("     {}. {}", i + 1, text);
    }

    // Generate embeddings for all texts
    let requests: Vec<EmbedContentRequest> = texts.iter().map(|text| {
        EmbedContentRequest {
            content: Content {
                role: None,
                parts: vec![
                    Part::Text {
                        text: text.to_string(),
                    }
                ],
            },
            task_type: Some(TaskType::SemanticSimilarity),
            title: None,
            output_dimensionality: None,
        }
    }).collect();

    let response = client.embeddings()
        .batch_embed("text-embedding-004", requests)
        .await?;

    if let Some(embeddings) = &response.embeddings {
        println!("\n   ✓ Embeddings generated");

        // Calculate cosine similarities
        if embeddings.len() >= 3 {
            let emb1 = &embeddings[0].values.as_ref().unwrap();
            let emb2 = &embeddings[1].values.as_ref().unwrap();
            let emb3 = &embeddings[2].values.as_ref().unwrap();

            let sim_1_2 = cosine_similarity(emb1, emb2);
            let sim_1_3 = cosine_similarity(emb1, emb3);
            let sim_2_3 = cosine_similarity(emb2, emb3);

            println!("\n   Cosine similarities:");
            println!("     Text 1 vs Text 2: {:.4}", sim_1_2);
            println!("     Text 1 vs Text 3: {:.4}", sim_1_3);
            println!("     Text 2 vs Text 3: {:.4}", sim_2_3);

            println!("\n   Analysis:");
            println!("     • Text 1 and 2 are semantically similar (same meaning)");
            println!("     • Text 3 is less similar (different topic)");
            println!("     • Higher cosine similarity = more similar meaning");
        }
    }

    println!("\n=== Embedding Use Cases ===");
    println!("  • Semantic search: Find relevant documents");
    println!("  • Recommendation systems: Find similar items");
    println!("  • Clustering: Group similar content");
    println!("  • Classification: Categorize text");
    println!("  • Duplicate detection: Find near-duplicates");

    Ok(())
}

/// Calculate cosine similarity between two vectors.
fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    if a.len() != b.len() {
        return 0.0;
    }

    let dot_product: f32 = a.iter().zip(b.iter()).map(|(x, y)| x * y).sum();
    let norm_a: f32 = a.iter().map(|x| x * x).sum::<f32>().sqrt();
    let norm_b: f32 = b.iter().map(|x| x * x).sum::<f32>().sqrt();

    if norm_a == 0.0 || norm_b == 0.0 {
        return 0.0;
    }

    dot_product / (norm_a * norm_b)
}
